chain/client/src/sync/handler.rs

```diff
@@ -0,0 +1,367 @@
+            highest_height,
+            &highest_height_peers,
+        );
+        unwrap_and_report_state_sync_result!(epoch_sync_result);
+
+        // Run header sync as long as there are headers to catch up.
+        let header_sync_result = self.header_sync.run(
+            &mut self.sync_status,
+            &chain,
+            highest_height,
+            &highest_height_peers,
+        );
+        unwrap_and_report_state_sync_result!(header_sync_result);
+        // Only body / state sync if header height is close to the latest.
+        let header_head = unwrap_and_report_state_sync_result!(chain.header_head());
+
+        // We should state sync if it's already started or if we have enough
+        // headers and blocks. The should_state_sync method may run block sync.
+        let should_state_sync =
+            self.should_state_sync(chain, header_head, highest_height, highest_height_peers);
+        let should_state_sync = unwrap_and_report_state_sync_result!(should_state_sync);
+        if !should_state_sync {
+            return None;
+        }
+        let update_sync_status_result = self.update_sync_status(chain);
+        unwrap_and_report_state_sync_result!(update_sync_status_result);
+
+        let sync_hash = match &self.sync_status {
+            SyncStatus::StateSync(s) => s.sync_hash,
+            // sync hash isn't known yet. Return and try again later.
+            _ => return None,
+        };
+
+        let me = signer.as_ref().map(|x| x.validator_id().clone());
+        let block_header = chain.get_block_header(&sync_hash);
+        let block_header = unwrap_and_report_state_sync_result!(block_header);
+        let shards_to_sync = get_shards_cares_about_this_or_next_epoch(
+            me.as_ref(),
+            true,
+            &block_header,
+            &shard_tracker,
+            chain.epoch_manager.as_ref(),
+        );
+
+        let blocks_to_request = self.request_sync_blocks(chain, block_header, highest_height_peers);
+        let blocks_to_request = unwrap_and_report_state_sync_result!(blocks_to_request);
+
+        let state_sync_status = match &mut self.sync_status {
+            SyncStatus::StateSync(s) => s,
+            _ => unreachable!("Sync status should have been StateSync!"),
+        };
+
+        // Waiting for all the sync blocks to be available because they are
+        // needed to finalize state sync.
+        if !blocks_to_request.is_empty() {
+            tracing::debug!(target: "sync", "waiting for sync blocks");
+            return Some(SyncHandlerRequest::NeedRequestBlocks(blocks_to_request));
+        }
+
+        let state_sync_result = self.state_sync.run(
+            sync_hash,
+            state_sync_status,
+            &highest_height_peers,
+            &shards_to_sync,
+        );
+        let state_sync_result = unwrap_and_report_state_sync_result!(state_sync_result);
+        match state_sync_result {
+            StateSyncResult::InProgress => return None,
```

Comment:
nit: maybe just `if ==` (might have to add `Eq, PartialEq`) instead of the old code's `match`